### DISCUSSION TITLE
API Rule: Target specific ID for multi-relation field

### DISCUSSION THREAD

**Discussion by jimafisk** [pocketbase/pocketbase#3104] [Q&A] [answered]
**Background:** We have Users who can join multiple Organizations. We keep track of the relations of Users to Organizations in a third collection (organization_relations) that has additional information such as the User's role for that particular Organization.

**Desired Outcome:** We want an API Rule that only lets Users of role "admin" for that particular Organization list records. 

**Question:** Is there a way to target a specific Organization ID and check the User's role for that ID only? I _think_ this might be what's happening automatically when using `&&` since you've [mentioned before](https://github.com/pocketbase/pocketbase/discussions/788#discussioncomment-3893650): 
> internally the rule is converted to a left join

**Current API Rule:**
```
organization_id ?= @request.auth.organization_relations.organization_id && @request.auth.organization_relations.roles ?~ "admin"
```

What I think we're saying ^ is allow listing records if:
1. one of our relations contains the ID
2. AND for that matched ID only, if at least one of our roles is "admin"

Is my understanding here correct? Thanks so much for the help!

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not entire sure that I understand the structure and where the relation fields reside.
Additionally if the above rule is supposed to be for the "organizations" collection then I guess that the "organization_id" field must be "id", right?

If you have a direct relation field in the "users" collection pointing to the "organization_relations" collection, then the above is correct.
One minor optimization if you want direct comparison and not `LIKE` could be, if "roles" in the above is a multiple `select` field, to use the [`:each` modifier](https://pocketbase.io/docs/api-rules-and-filters/#each-modifier), aka.:

```js
id ?= @request.auth.organization_relations.organization_id &&
@request.auth.organization_relations.roles:each ?= "admin"
```

The above will be roughly equivalent to the following query:

```sql
-- "HAS" is a fake operator for simplicity (internally json_each is used)
SELECT org.*
FROM organizations org
LEFT JOIN users on users.id = {:authId}
LEFT JOIN organization_relations rels on users.organization_relations HAS rels.id
WHERE (
    org.id = rels.organization_id AND
    rels.roles HAS "admin"
)
```

--- BUT ---

If you **don't have direct relation** field in the "users" collection pointing to "organization_relations" and instead there is a relation field in the latter pointing to the "users", then the above will not work because the user record will not have "organization_relations" field.
In this case, what you want seems to be access to an indirect relation, but this currently is not supported directly. For this purpose you can use the `@collection.*` filter to join the 3rd collection. For example (_assuming that there is a "user_id" field inside "organization_relations"_):

```js
@collection.organization_relations.organization_id ?= id &&
@collection.organization_relations.user_id ?= @request.auth.id &&
@collection.organization_relations.roles:each ?= "admin"
```

The above will be roughly equivalent to the following query:

```sql
-- "HAS" is a fake operator for simplicity (internally json_each is used)
SELECT org.*
FROM organizations org
LEFT JOIN organization_relations rels
WHERE (
    rels.organization_id = org.id AND
    rels.user_id = {:authId} AND
    rels.roles HAS "admin"
)
```

_If the above is not what you are looking for, please export your collections structure (or only portion of it) as JSON (_Admin UI > Settings > Export collections_) and elaborate for which collection the API rule is defined. I'll consider in the future to embed an ERD export/view mode in the Admin UI to make debugging and reporting slightly easier._

**Reply by jimafisk** [reply]
Wow such a thorough answer (as always), thank you @ganigeorgiev! This gives me exactly what I needed (and more) :tada: 

I'm sorry for the vague parts in my original post, I was trying to simplify things but I see how I introduced confusion.

> Additionally if the above rule is supposed to be for the "organizations" collection then I guess that the "organization_id" field must be "id", right?

In reality I was working with a View collection, but you're 100% right.

> If you have a direct relation field in the "users" collection pointing to the "organization_relations" collection, then the above is correct.

Great thank you! Yes, my intent was there was a relation field on the User with the same name as the collection it was relating to (e.g. `organization_relations`)

> One minor optimization if you want direct comparison and not LIKE could be, if "roles" in the above is a multiple select field, to use the [:each modifier](https://pocketbase.io/docs/api-rules-and-filters/#each-modifier)

Good catch, I assume the `?~` would match "admin" but also "superadmin", whereas `:each ?=` will only match "admin". Perfect :ok_hand: 

> For this purpose you can use the @collection.* filter to join the 3rd collection.

Very helpful to know this for the future, thank you!
